### ISSUE TITLE
Feature request: Periodic boundary conditions ("wrap" mode) for ndimage morphology functions (dilation, erosion, opening, closing)

### ISSUE THREAD

**Issue by smheidrich** [scipy/scipy#8218]
Most of the important functions in [``scipy.ndimage``](https://docs.scipy.org/doc/scipy-1.0.0/reference/ndimage.html) support periodic boundary conditions via ``mode="wrap"`` (e.g. [convolve](https://docs.scipy.org/doc/scipy-1.0.0/reference/generated/scipy.ndimage.convolve.html)), but this does not seem to be the case for any of the morphology functions, such as [binary_erosion](https://docs.scipy.org/doc/scipy-1.0.0/reference/generated/scipy.ndimage.binary_erosion.html), [binary_dilation](https://docs.scipy.org/doc/scipy-1.0.0/reference/generated/scipy.ndimage.binary_dilation.html), [binary_opening](https://docs.scipy.org/doc/scipy-1.0.0/reference/generated/scipy.ndimage.binary_opening.html) or [binary_closing](https://docs.scipy.org/doc/scipy-1.0.0/reference/generated/scipy.ndimage.binary_closing.html).

It would be nice if periodic boundary conditions were implemented for those as well.

As far as I can tell, their [equivalent in OpenCV](https://docs.opencv.org/2.4/modules/imgproc/doc/filtering.html#Ptr%3CFilterEngine%3E%20createMorphologyFilter%28int%20op,%20int%20type,%20InputArray%20kernel,%20Point%20anchor,%20int%20rowBorderType,%20int%20columnBorderType,%20const%20Scalar&%20borderValue%29) does have this functionality (via ``BORDER_WRAP``), although I haven't tested it. Let me know if what I'm asking doesn't make any sense.

**Comment by harrymd** [comment]
This would also be useful for ndimage.label().

**Comment by rabernat** [comment]
This feature is very important to our group. Would the maintainers be open to us contributing a pull request to implement this enhancement?

**Comment by tylerjereddy** [comment]
some relevant experts/maintainers maybe: @grlee77 @stefanv

**Comment by stefanv** [comment]
That seems like a straightforward extension, so I see no reason why not.  /cc @mkcor @jni @sciunto @rfezzani @emmanuelle

**Comment by rabernat** [comment]
After our initial enthusiasm, our group spent quite a bit of time looking at [_ni_label.pyx](https://github.com/scipy/scipy/blob/master/scipy/ndimage/src/_ni_label.pyx) (the `label` function is our most urgent need).

Unfortunately we are having a really hard time understanding this code. We have written some cython before, but this feels complex beyond our pay grade.

We were was hoping to find the place in this code where the edge of the array is handled, and then add a special treatment for periodic boundaries. My best guess is that it is somewhere in this function
https://github.com/scipy/scipy/blob/505a8fc6cf160562830c102919e0d6d42e0a7365/scipy/ndimage/src/_ni_label.pyx#L168-L177

If anyone familiar with this code could give us some hints, it would be super helpful. It looks like most of this code is about 9 years old, so there may not be so much institutional memory here. But it was touched somewhat recently by @WarrenWeckesser as part of some code maintenance...

Without some guidance, I'm afraid we may have to back down from our overly ambitious goals.

**Comment by grlee77** [comment]
I will try to take a look. I have never worked on these morphology functions here, but am familiar with how it is done for the `filters` and `interpolation` components of scikit-image.

**Comment by grlee77** [comment]
For the binary morphology operations, they are all built based on a single binary erosion function in the C code. That function looks fairly similar in design to the functions for the convolution-based filters. It is just that the mode is always set to constant boundary extension and the user is only able to choose whether that constant is 0 or 1.

Specifically, where the code here for correlation uses a `mode` parameter:
https://github.com/scipy/scipy/blob/505a8fc6cf160562830c102919e0d6d42e0a7365/scipy/ndimage/src/ni_filters.c#L208-L212
For binary erosion it is hardcoded `NI_EXTEND_CONSTANT`
https://github.com/scipy/scipy/blob/505a8fc6cf160562830c102919e0d6d42e0a7365/scipy/ndimage/src/ni_morphology.c#L114-L119

`label` is unique in that it is a separate Cython implementation of an algorithm I am not familiar with.

For the case of label, specifically, I wonder if it might not be easier to support `wrap` mode specifically via relabeling after using the existing implementation. e.g. for a trivial example, the following set of labels:

[1 1 1 0 0 0 2 2 2 0 0 0 3 3 3] would become [1 1 1 0 0 0 2 2 2 0 0 0 1 1 1] because labels 1 and 3 would be part of the same object under periodic conditions?

**Comment by rabernat** [comment]
> [1 1 1 0 0 0 2 2 2 0 0 0 3 3 3] would become [1 1 1 0 0 0 2 2 2 0 0 0 1 1 1] because labels 1 and 3 would be part of the same object under periodic conditions?

Good idea. But one challenge is that you might have to reindex all of your features, for example, imagine the scenario where the same features are indexed differently:  [1 1 1 0 0 0 3 3 3 0 0 0 2 2 2] -->  [1 1 1 0 0 0 3 3 3 0 0 0 1 1 1]. Now we are missing the label 2, so we would really want to have [1 1 1 0 0 0 2 2 2 0 0 0 1 1 1], i.e. relabeling the middle feature as well. Could get tricky.

**Comment by grlee77** [comment]
> Good idea. But one challenge is that you might have to reindex all of your features

Yes, but fortunately there is already a [relabel_sequential](https://scikit-image.org/docs/dev/api/skimage.segmentation.html#skimage.segmentation.relabel_sequential) utility in scikit-image that we could borrow from

**Comment by rfezzani** [comment]
The `return_inverse` argument of `np.unique` can also do the trick :wink:

```python
>>> a = zeros((10, 10), int)
>>> a[:2, :2] = 1
>>> a[5:7, 3:8] = 3
>>> a
array([[1, 1, 0, 0, 0, 0, 0, 0, 0, 0],
       [1, 1, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 3, 3, 3, 3, 3, 0, 0],
       [0, 0, 0, 3, 3, 3, 3, 3, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]])
>>> np.unique(a, return_inverse=True)[1].reshape(a.shape)
array([[1, 1, 0, 0, 0, 0, 0, 0, 0, 0],
       [1, 1, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 2, 2, 2, 2, 2, 0, 0],
       [0, 0, 0, 2, 2, 2, 2, 2, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]])

```
